google/resource_storage_object_acl.go

```diff
@@ -56,196 +102,247 @@ func resourceStorageObjectAclCreate(d *schema.ResourceData, meta interface{}) er
-				log.Printf("[DEBUG]: saving re %s-%s", role, entity)
-			}
+		err = performStorageObjectRoleEntityOperations(create, update, remove, config, bucket, object)
+		if err != nil {
+			return fmt.Errorf("Error updating object %s in %s: %v", object, bucket, err)
 		}
 
-		d.Set("role_entity", role_entity)
-	} else {
-		d.Set("role_entity", nil)
+		return resourceStorageObjectAclRead(d, meta)
 	}
-
-	d.SetId(getObjectAclId(object))
-	return nil
 }
 
-func resourceStorageObjectAclUpdate(d *schema.ResourceData, meta interface{}) error {
+func resourceStorageObjectAclDelete(d *schema.ResourceData, meta interface{}) error {
 	config := meta.(*Config)
 
 	bucket := d.Get("bucket").(string)
 	object := d.Get("object").(string)
 
-	if d.HasChange("role_entity") {
-		o, n := d.GetChange("role_entity")
-		old_re, new_re := o.([]interface{}), n.([]interface{})
-
-		old_re_map := make(map[string]string)
-		for _, v := range old_re {
-			res, err := getRoleEntityPair(v.(string))
+	res, err := config.clientStorage.Objects.Get(bucket, object).Do()
+	if err != nil {
+		return fmt.Errorf("Error reading object %s in %s: %v", object, bucket, err)
+	}
 
-			if err != nil {
-				return fmt.Errorf(
-					"Old state has malformed Role/Entity pair: %v", err)
-			}
+	res, err = config.clientStorage.Objects.Update(bucket, object, res).PredefinedAcl("private").Do()
+	if err != nil {
+		return fmt.Errorf("Error updating object %s in %s: %v", object, bucket, err)
+	}
 
-			old_re_map[res.Entity] = res.Role
-		}
+	return nil
+}
 
-		for _, v := range new_re {
-			pair, err := getRoleEntityPair(v.(string))
+func getRoleEntitiesAsStringsFromApi(config *Config, bucket string, object string) ([]string, error) {
+	var roleEntities []string
+	res, err := config.clientStorage.ObjectAccessControls.List(bucket, object).Do()
+	if err != nil {
+		return nil, err
+	}
 
-			objectAccessControl := &storage.ObjectAccessControl{
-				Role:   pair.Role,
-				Entity: pair.Entity,
-			}
+	for _, roleEntity := range res.Items {
+		role := roleEntity.Role
+		entity := roleEntity.Entity
+		roleEntities = append(roleEntities, fmt.Sprintf("%s:%s", role, entity))
+	}
 
-			// If the old state is missing this entity, it needs to
-			// be created. Otherwise it is updated
-			if _, ok := old_re_map[pair.Entity]; ok {
-				_, err = config.clientStorage.ObjectAccessControls.Update(
-					bucket, object, pair.Entity, objectAccessControl).Do()
-			} else {
-				_, err = config.clientStorage.ObjectAccessControls.Insert(
-					bucket, object, objectAccessControl).Do()
-			}
+	return roleEntities, nil
+}
 
-			// Now we only store the keys that have to be removed
-			delete(old_re_map, pair.Entity)
+// Creates 3 lists of changes we need to make to go from one set of entities to another- which entities need to be created, update, and deleted
+// Not resource specific
+func getRoleEntityChange(old []interface{}, new []interface{}, owner string) ([]*RoleEntity, []*RoleEntity, []*RoleEntity, error) {
```

Comment:
if you make old and new []strings, I think you can get rid of some code that's converting them from strings -> interfaces and back